### PR TITLE
Changed copyright check to the new standard from PR 1422.

### PR DESCRIPTION
**Pull request by YoshikiTakashima**
### Description of changes: 

Changes the behavior of the copyright check to allow for licenses that are not at the top (external code license policy given in PR1422).

### Resolved issues:

Adaptation to #1422

### Call-outs:

Copyright add

### Testing:

* How is this change tested? Modifications to  the following files trigger changed functions
- tests/cargo-kani/vecdeque-cve/src/cve.rs
- tests/cargo-kani/vecdeque-cve/src/fixed.rs
- tests/cargo-kani/vecdeque-cve/src/raw_vec.rs

* Is this a refactor change? No

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by YoshikiTakashima** [comment]
This PR is blocking #1435. Need merge.

File: scripts/ci/copyright_check.py [hunk 1]
```diff
@@ -28,6 +28,17 @@ def get_header(has_shebang, regexes):
...
 
+
+def match_somewhere(regexes, lines):
+    maybe_match_head_index = [index for index, line in enumerate(lines) if regexes[0].search(line)][:1]
+    if maybe_match_head_index and maybe_match_head_index[0] + len(regexes) <= len(lines):
+        match_head_index = maybe_match_head_index[0]
+        matches = [regex.search(lines[match_head_index + index]) for index, regex in enumerate(regexes)]
+        return all(matches)
+    else:
+        print('error return')
```
Comment (adpaco-aws) [inline scripts/ci/copyright_check.py:39]: This shouldn't be printed.

File: scripts/ci/copyright_check.py [hunk 2]
```diff
@@ -28,6 +28,17 @@ def get_header(has_shebang, regexes):
     indices = range(init_idx, init_idx + len(regexes))
     return zip(regexes, indices)
 
+
+def match_somewhere(regexes, lines):
```
Comment (adpaco-aws) [inline scripts/ci/copyright_check.py:32]: Can you please add a brief description?

**Review by adpaco-aws — COMMENTED** [review]
Thanks! Two comments.

File: scripts/ci/copyright_check.py (hunk 1 above)
Comment (YoshikiTakashima) [reply scripts/ci/copyright_check.py:39]: Thanks. Forgot to get rid of it after debugging.

Comment (YoshikiTakashima) [reply scripts/ci/copyright_check.py:39]: Fixed with 4c48d334d4fd1631529108a23ee127754bab87be

File: scripts/ci/copyright_check.py (hunk 2 above)
Comment (YoshikiTakashima) [reply scripts/ci/copyright_check.py:32]: Fixed with df6e3a0244c78c2376c35d27e9714834077b9db0

File: scripts/ci/copyright_check.py [hunk 3]
```diff
@@ -28,6 +28,18 @@ def get_header(has_shebang, regexes):
     indices = range(init_idx, init_idx + len(regexes))
     return zip(regexes, indices)
 
+# Matches all MODIFIED_HEADER patterns within the file. This is used
+# when the license is not at the top header, and there are licenses of
+# external libraries.
+def match_somewhere(regexes, lines):
+    maybe_match_head_index = [index for index, line in enumerate(lines) if regexes[0].search(line)][:1]
```
Comment (celinval) [inline scripts/ci/copyright_check.py:35]: Will this match adding the copyright header in the middle of the file?

File: tests/cargo-kani/vecdeque-cve/src/fixed.rs [hunk 4]
```diff
@@ -1,3 +1,5 @@
+// From Rust Standard Library. Copyright Rust Developers.
```
Comment (celinval) [inline tests/cargo-kani/vecdeque-cve/src/fixed.rs:1]: I don't think rust std uses copyright statements.

File: scripts/ci/copyright_check.py (hunk 3 above)
Comment (YoshikiTakashima) [reply scripts/ci/copyright_check.py:35]: Yes it does. I’m not sure if there’s a good way to do check if our license is directly below the other license, but I can try to assert that all lines above start with a comment or is empty. Let me know if that works.

File: tests/cargo-kani/vecdeque-cve/src/fixed.rs (hunk 4 above)
Comment (YoshikiTakashima) [reply tests/cargo-kani/vecdeque-cve/src/fixed.rs:1]: Artificially added for testing purposes.

File: scripts/ci/copyright_check.py [hunk 5]
```diff
@@ -28,6 +28,18 @@ def get_header(has_shebang, regexes):
     indices = range(init_idx, init_idx + len(regexes))
     return zip(regexes, indices)
 
+# Matches all MODIFIED_HEADER patterns within the file. This is used
```
[the remark this answers is not in the document]
Comment (YoshikiTakashima) [reply scripts/ci/copyright_check.py]: Fixed with df6e3a0244c78c2376c35d27e9714834077b9db0

File: tests/cargo-kani/vecdeque-cve/src/fixed.rs (hunk 4 above)
Comment (YoshikiTakashima) [reply tests/cargo-kani/vecdeque-cve/src/fixed.rs]: @fzaiser's generator test cases already had injected copyright notices, so I moved up the licenses them so they test the copyright script. They are merged in main, so I feel it's ok to use.

File: scripts/ci/copyright_check.py (hunk 3 above)
Comment (YoshikiTakashima) [reply scripts/ci/copyright_check.py:37]: Implemented proposal with https://github.com/model-checking/kani/pull/1447/commits/f4665b06a3fde09a1a496f35fc8e663dff296cfa

Feel free to reopen if it is not what you intended.

File: scripts/ci/copyright_check.py [hunk 6]
```diff
@@ -28,6 +30,19 @@ def get_header(has_shebang, regexes):
...
     return zip(regexes, indices)
 
+def match_somewhere(regexes, lines, empty_or_comment_regex):
+    """ Matches all MODIFIED_HEADER patterns within the file. This is used
+    when the license is not at the top header, and there are licenses of
+    external libraries. """
+    maybe_match_head_index = [index for index, line in enumerate(lines) if regexes[0].search(line)][:1]
+    if maybe_match_head_index and maybe_match_head_index[0] + len(regexes) <= len(lines):
+        match_head_index = maybe_match_head_index[0]
+        matches_pre_license = [empty_or_comment_regex.search(lines[index]) for index in range(match_head_index)]
```
Comment (celinval) [inline scripts/ci/copyright_check.py:40]: nit: prefer using "()" instead of "[]". This use case is not that bad since it's creating a list of boolean.

**Review by celinval — APPROVED** [review]
Thank you for doing this.

File: scripts/ci/copyright_check.py (hunk 6 above)
Comment (YoshikiTakashima) [reply scripts/ci/copyright_check.py]: https://github.com/model-checking/kani/pull/1447/commits/361a8975a28dad766354549cb876697e584821b4

**Pull request merged**